src/CsvHelper/CsvParser.cs

```diff
@@ -281,15 +369,56 @@ public async Task<bool> ReadAsync()
 			}
 		}
 
+		private async Task<bool> FillBufferAsync()

```

Comment:
1. You have fully repeated the FillBuffer method's code, it's not good. Don't you think it would be better to call FillBufferAsync().Result from the FillBuffer method?
2. The method's purpose is not clear, because of a lot of global fields. I don't think it's a good idea to refactor this class now, so add a comment line above both FillBuffer and FillBufferAsync
3. It's a little unclear what is going on in lines 386 to 407, maybe we could call an additional method with a good name?